### PR TITLE
GPMSA normalization options for output data

### PR DESCRIPTION
**Pull request by roystgnr**
This adds analogous scaling options to what we support for the input data, implements output autoscaling calculations within GPMSAOptions, and implements output scaling within GPMSA.

The "autoscale-all" regression test results changed, as expected, which unfortunately does not necessarily prove the results changed as expected.

### CODE REVIEW

File: src/gp/inc/GPMSAOptions.h [hunk 1]
```diff
@@ -100,6 +100,12 @@ class GPMSAOptions
   void set_autoscale_minmax_scenario_parameter(unsigned int i);
 
   //! Do automatic normalization, using minimum and maximum values in
+  //  the supplied data, for output data i.
+  //
+  //  Normalized values will range from 0 to 1.
+  void set_autoscale_minmax_output(unsigned int i);
```
Comment (briadam) [inline src/gp/inc/GPMSAOptions.h]: Good to have this for completeness, though Brian Williams recommends always using mean/variance scaling.

File: src/gp/inc/GPMSAOptions.h [hunk 2]
```diff
@@ -119,6 +125,12 @@ class GPMSAOptions
   void set_autoscale_meanvar_scenario_parameter(unsigned int i);
 
   //! Do automatic normalization, using mean and variance of the
+  //  supplied data, for output i.
+  //
+  //  Normalized values will have mean 0 and standard deviation 1.
+  void set_autoscale_meanvar_output(unsigned int i);
```
Comment (briadam) [inline src/gp/inc/GPMSAOptions.h]: As we think forward to the functional case, probably don't need scaling on a per-component basis.  More likely on a per-group basis, but having this generality is fine.

File: src/gp/src/GPMSA.C [hunk 3]
```diff
@@ -1155,6 +1160,8 @@ GPMSAFactory<V, M>::setUpHyperpriors()
       V y_temp(Wy * y);
       y_temp -= Wy * B * yhat;
 
+      // At this point everything has already been normalized, so no
+      // more normalization to do?
```
Comment (briadam) [inline src/gp/src/GPMSA.C]: I'm a little surprised that we don't have to scale the experiment covariance information provided by the user, as we did in examples/gp/linear_verif/  Or maybe you are with Wy above...

File: src/gp/src/GPMSAOptions.C [hunk 4]
```diff
@@ -558,6 +594,30 @@ GPMSAOptions::set_final_scaling
...
+      for (unsigned int i=1; i < m_simulationOutputs.size(); ++i)
+        min_max_update(minOutput, maxOutput,
+                       (*m_simulationOutputs[i]));
+
+      for (unsigned int p=0; p != dimOutput; ++p)
+        if (m_autoscaleMinMaxAll ||
+            m_autoscaleMinMaxOutput.count(p))
+          {
+            if ((m_outputScaleMin.size() > p) &&
+                ((m_outputScaleMin[p] != 0) ||
```
Comment (briadam) [inline src/gp/src/GPMSAOptions.C]: Potentially dangerous test for floating point equality.

File: src/gp/src/GPMSAOptions.C [hunk 5]
```diff
@@ -624,6 +684,36 @@ GPMSAOptions::set_final_scaling
...
+                        *m_simulationOutputs[i]);
+
+      varOutput /= n;
+
+      for (unsigned int p=0; p != dimOutput; ++p)
+        if (m_autoscaleMeanVarAll ||
+            m_autoscaleMeanVarOutput.count(p))
+          {
+            if ((m_outputScaleMin.size() > p) &&
+                ((m_outputScaleMin[p] != 0) ||
```
Comment (briadam) [inline src/gp/src/GPMSAOptions.C]: Potentially dangerous test for floating point equality.

File: src/gp/src/GPMSA.C [hunk 6]
```diff
@@ -766,7 +767,8 @@ GPMSAFactory<V, M>::setUpEmulator()
 
   for (unsigned int i=0; i != m_numSimulations; ++i)
     for (unsigned int j=0; j != numOutputs; ++j)
-      (*simulationOutputMeans)[j] += (*m_simulationOutputs[i])[j];
+      (*simulationOutputMeans)[j] +=
```
Comment (briadam) [inline src/gp/src/GPMSA.C]: Verify that this isn't doubly applying the mean shift between the code here and in normalized_output.

File: src/gp/src/GPMSA.C [hunk 7]
```diff
@@ -776,7 +778,8 @@ GPMSAFactory<V, M>::setUpEmulator()
   for (unsigned int i=0; i != m_numSimulations; ++i)
     for (unsigned int j=0; j != numOutputs; ++j)
       simulation_matrix(i,j) =
-        (*m_simulationOutputs[i])[j] - (*simulationOutputMeans)[j];
+        this->m_opts->normalized_output(j, (*m_simulationOutputs[i])[j]) -
+        (*simulationOutputMeans)[j];
```
Comment (briadam) [inline src/gp/src/GPMSA.C]: Verify that this isn't doubly applying the mean shift between the code here and in normalized_output.  (And in all places normalized_output is being called below.)

File: src/gp/src/GPMSA.C [hunk 8]
```diff
@@ -904,7 +907,9 @@ GPMSAFactory<V, M>::setUpEmulator()
               // No fancy perturbation here
               unsigned int j = ex*numOutputs+outj;
 
-              Wy(i,j) = W_i(outi,outj);
+              Wy(i,j) = W_i(outi,outj) *
```
Comment (briadam) [inline src/gp/src/GPMSA.C]: Perhaps this is accounting for the need to scale the user's observation covariance?

**Review by briadam — CHANGES_REQUESTED** [review]
Change might not strictly be needed, but comments should be reviewed before merge.

File: src/gp/inc/GPMSAOptions.h (hunk 2 above)
Comment (roystgnr) [inline src/gp/inc/GPMSAOptions.h]: Yeah; I think when we add the partitioning capability, this won't be "output i" but rather "output subpartition i".

File: src/gp/src/GPMSA.C (hunk 6 above)
Comment (roystgnr) [inline src/gp/src/GPMSA.C]: This code is correct, but it's subtle.  If you're doing mean/variance normalization on a single index i, then simulationOutputMeans will be 0, and calculating it is a waste of time, but still gives the correct result.

But, if you're normalizing based on min/max, or a user-submitted range, or if index i is only one index of many in a partition, then the mean of the normalized values will not be zero, and yet we'll still want to subtract it off before doing the SVD.

File: src/gp/src/GPMSA.C (hunk 8 above)
Comment (roystgnr) [inline src/gp/src/GPMSA.C]: This was a bug, fixed in the latest commit; thanks.

File: src/gp/src/GPMSA.C (hunk 3 above)
Comment (roystgnr) [inline src/gp/src/GPMSA.C]: I was scaling it, but in the *wrong* direction, before that fix.

File: src/gp/src/GPMSAOptions.C (hunk 4 above)
Comment (roystgnr) [inline src/gp/src/GPMSAOptions.C]: Nah.  The fill in set_output_scaling is with *exactly* 0 and 1, so we can rely on an exact test to tell us if something other than a fill happened.

Now, there is a loophole here: if someone manually sets 0 and 1, then requests automatic scaling, then we won't catch that and scream at them - by doing something insane *twice* they'll have evaded our sanity check...

**Review by briadam — APPROVED** [review]

**Pull request merged**